Use CMake Targets

Fix buildpaths QA errors by backporting selected CMake changes to
use CMake targets.

Upstream-Status: Backport [https://github.com/ros-navigation/navigation2/commit/b6a5387d6823de6eb8008da995a72086e8ace1e5]

Signed-off-by: Rob Woolley <rob.woolley@windriver.com>

Index: nav2-route-1.1.20-1/CMakeLists.txt
===================================================================
--- nav2-route-1.1.20-1.orig/CMakeLists.txt
+++ nav2-route-1.1.20-1/CMakeLists.txt
@@ -29,22 +29,6 @@ include_directories(
 set(executable_name route_server)
 set(library_name ${executable_name}_core)
 
-set(dependencies
-  rclcpp
-  rclcpp_lifecycle
-  rclcpp_components
-  std_msgs
-  geometry_msgs
-  nav2_costmap_2d
-  pluginlib
-  visualization_msgs
-  nav_msgs
-  tf2_ros
-  nav2_core
-  nanoflann
-  nlohmann_json
-)
-
 # Main library
 add_library(${library_name} SHARED
   src/route_server.cpp
@@ -63,12 +47,40 @@ ament_target_dependencies(${library_name
   ${dependencies}
 )
 
+target_include_directories(${library_name}
+  PUBLIC
+    "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
+    "$<INSTALL_INTERFACE:include/${PROJECT_NAME}>"
+)
+target_link_libraries(${library_name} PUBLIC
+  nav2_core::nav2_core
+  nav2_costmap_2d::nav2_costmap_2d_core
+  ${nav2_msgs_TARGETS}
+  nav2_util::nav2_util_core
+  ${nav_msgs_TARGETS}
+  pluginlib::pluginlib
+  rclcpp::rclcpp
+  rclcpp_lifecycle::rclcpp_lifecycle
+  ${rcl_interfaces_TARGETS}
+  ${geometry_msgs_TARGETS}
+  ${visualization_msgs_TARGETS}
+  ${std_msgs_TARGETS}
+  ${nav_msgs_TARGETS}
+  tf2_ros::tf2_ros
+  nlohmann_json::nlohmann_json
+)
+target_link_libraries(${library_name} PRIVATE
+  ${lifecycle_msgs_TARGETS}
+  rclcpp_components::component
+  tf2::tf2
+)
+
 # Main executable
 add_executable(${executable_name}
   src/main.cpp
 )
 
-target_link_libraries(${executable_name} ${library_name})
+target_link_libraries(${executable_name} PRIVATE ${library_name} rclcpp::rclcpp)
 
 ament_target_dependencies(${executable_name}
   ${dependencies}
@@ -88,8 +100,33 @@ add_library(edge_scorers SHARED
   src/plugins/edge_cost_functions/start_pose_orientation_scorer.cpp
 )
 
-ament_target_dependencies(edge_scorers
-  ${dependencies}
+target_include_directories(edge_scorers
+  PUBLIC
+    "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
+    "$<INSTALL_INTERFACE:include/${PROJECT_NAME}>"
+)
+target_link_libraries(edge_scorers PUBLIC
+  nav2_core::nav2_core
+  nav2_costmap_2d::nav2_costmap_2d_core
+  ${nav2_msgs_TARGETS}
+  nav2_util::nav2_util_core
+  ${nav_msgs_TARGETS}
+  pluginlib::pluginlib
+  rclcpp::rclcpp
+  rclcpp_lifecycle::rclcpp_lifecycle
+  ${rcl_interfaces_TARGETS}
+  ${geometry_msgs_TARGETS}
+  ${visualization_msgs_TARGETS}
+  ${std_msgs_TARGETS}
+  ${nav_msgs_TARGETS}
+  tf2_ros::tf2_ros
+  angles::angles
+  nlohmann_json::nlohmann_json
+)
+target_link_libraries(edge_scorers PRIVATE
+  ${lifecycle_msgs_TARGETS}
+  rclcpp_components::component
+  tf2::tf2
 )
 
 # Route operations plugins
@@ -101,8 +138,32 @@ add_library(route_operations SHARED
   src/plugins/route_operations/time_marker.cpp
 )
 
-ament_target_dependencies(route_operations
-  ${dependencies}
+target_include_directories(route_operations
+  PUBLIC
+    "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
+    "$<INSTALL_INTERFACE:include/${PROJECT_NAME}>"
+)
+target_link_libraries(route_operations PUBLIC
+  nav2_core::nav2_core
+  nav2_costmap_2d::nav2_costmap_2d_core
+  ${nav2_msgs_TARGETS}
+  nav2_util::nav2_util_core
+  ${nav_msgs_TARGETS}
+  pluginlib::pluginlib
+  rclcpp::rclcpp
+  rclcpp_lifecycle::rclcpp_lifecycle
+  ${rcl_interfaces_TARGETS}
+  ${geometry_msgs_TARGETS}
+  ${visualization_msgs_TARGETS}
+  ${std_msgs_TARGETS}
+  ${nav_msgs_TARGETS}
+  tf2_ros::tf2_ros
+  nlohmann_json::nlohmann_json
+)
+target_link_libraries(route_operations PRIVATE
+  ${lifecycle_msgs_TARGETS}
+  rclcpp_components::component
+  tf2::tf2
 )
 
 # Graph Parser plugins
@@ -110,16 +171,64 @@ add_library(graph_file_loaders SHARED
     src/plugins/graph_file_loaders/geojson_graph_file_loader.cpp
 )
 
-ament_target_dependencies(graph_file_loaders
-    ${dependencies}
+target_include_directories(graph_file_loaders
+  PUBLIC
+    "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
+    "$<INSTALL_INTERFACE:include/${PROJECT_NAME}>"
+)
+target_link_libraries(graph_file_loaders PUBLIC
+  nav2_core::nav2_core
+  nav2_costmap_2d::nav2_costmap_2d_core
+  ${nav2_msgs_TARGETS}
+  nav2_util::nav2_util_core
+  ${nav_msgs_TARGETS}
+  pluginlib::pluginlib
+  rclcpp::rclcpp
+  rclcpp_lifecycle::rclcpp_lifecycle
+  ${rcl_interfaces_TARGETS}
+  ${geometry_msgs_TARGETS}
+  ${visualization_msgs_TARGETS}
+  ${std_msgs_TARGETS}
+  ${nav_msgs_TARGETS}
+  tf2_ros::tf2_ros
+  nlohmann_json::nlohmann_json
+)
+target_link_libraries(graph_file_loaders PRIVATE
+  ${lifecycle_msgs_TARGETS}
+  rclcpp_components::component
+  tf2::tf2
 )
 
 add_library(graph_file_savers SHARED
     src/plugins/graph_file_savers/geojson_graph_file_saver.cpp
 )
 
-ament_target_dependencies(graph_file_savers
-    ${dependencies}
+target_include_directories(graph_file_savers
+  PUBLIC
+    "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
+    "$<INSTALL_INTERFACE:include/${PROJECT_NAME}>"
+)
+target_link_libraries(graph_file_savers PUBLIC
+  nav2_core::nav2_core
+  nav2_costmap_2d::nav2_costmap_2d_core
+  ${nav2_msgs_TARGETS}
+  nav2_util::nav2_util_core
+  ${nav_msgs_TARGETS}
+  pluginlib::pluginlib
+  rclcpp::rclcpp
+  rclcpp_lifecycle::rclcpp_lifecycle
+  ${rcl_interfaces_TARGETS}
+  ${geometry_msgs_TARGETS}
+  ${visualization_msgs_TARGETS}
+  ${std_msgs_TARGETS}
+  ${nav_msgs_TARGETS}
+  tf2_ros::tf2_ros
+  nlohmann_json::nlohmann_json
+)
+target_link_libraries(graph_file_savers PRIVATE
+  ${lifecycle_msgs_TARGETS}
+  rclcpp_components::component
+  tf2::tf2
 )
 
 pluginlib_export_plugin_description_file(nav2_route plugins.xml)
@@ -151,6 +260,20 @@ if(BUILD_TESTING)
 endif()
 
 ament_export_include_directories(include)
-ament_export_dependencies(${dependencies})
+ament_export_dependencies(
+  rclcpp
+  rclcpp_lifecycle
+  rclcpp_components
+  std_msgs
+  geometry_msgs
+  nav2_costmap_2d
+  pluginlib
+  visualization_msgs
+  nav_msgs
+  tf2_ros
+  nav2_core
+  nanoflann
+  nlohmann_json
+)
 ament_export_libraries(${library_name} edge_scorers route_operations graph_file_loaders)
 ament_package()
